### PR TITLE
Prevent infinite loop when data store fails

### PR DESCRIPTION
**Pull request by chlunde**
If we get an empty or short response, the process will loop forever trying to read the rest of the data.  Prevent this by raising an exception if the response is not long enough.  I saw this issue on other nodes when a data server ran out of file descriptors.

An alternative would be to make sure _ReadReply returns None for a more graceful retry.  What do you think?

I guess this is not a common issue, because processes stuck at 100% CPU forever tends to stick out.

### CODE REVIEW

**Comment by chlunde** [comment]
I believe there's also a similar potential problem two places in `grr/server/data_server/data_server.py`.

**Comment by chlunde** [comment]
I signed it!

File: grr/lib/data_stores/http_data_store.py [hunk 1]
```diff
@@ -83,7 +83,10 @@ def _ReadExactly(self, n):
     ret = ""
     left = n
     while left:
-      ret += self.sock.recv(left)
+      data = self.sock.recv(left)
+      if data == "":
+        raise IOError("Expected %d bytes, got EOF after %d", n, len(ret))
```
Comment (destijl) [inline grr/lib/data_stores/http_data_store.py]: maybe it makes more sense to raise socket.error so it will return none in readreply. @grrrrrrrrr you know this code better than me, can you take a look?

Comment (chlunde) [inline grr/lib/data_stores/http_data_store.py]: Note to self: I should also have used `% (n, len(ret))`.

Comment (grrrrrrrrr) [inline grr/lib/data_stores/http_data_store.py]: I agree that catching the exception would be better but I also like IOError better than socket error - the socket probably works just fine. Can you maybe add IOError to the exceptions we catch in _Reconnect and _ReadReply?

Comment (destijl) [inline grr/lib/data_stores/http_data_store.py]: And you're right about the format string, that needs fixing.

**Comment by chlunde** [comment]
Thanks for the review! I have updated the PR. I have not tested it because of the missing clients/#394.  I also removed `if not replylen_str` because I don't think it is - or was - reachable.

**Comment by destijl** [comment]
Thanks, can you give this a test now the clients are out? It's a fair bit of work for us to stand up a datastore just to test this.

**Comment by chlunde** [comment]
Sorry for the delay, I've been busy with other tasks.

I believe I found the root cause: Either through out of memory or too many open files on the data nodes four C.nnnnn.sqlite files were malformed.  This made the data server return an empty response, which caused this infinite loop.

With this patch on 0.3.0.post7, the system behaved better but there were still many issues and restarts were required.  After I removed the four corrupt client databases everything seems much better.  I believe there might be a few issues still left if this everything should work "smoothly" with a corrupt file:

* In sqlite_data_store.py, Execute tries to log the file name for a malformed database (#279), but in my case, the error was actually thrown later, for example in fetchall(), so the file name was never logged.
* I've seen "bad filedescriptor" errors and spinning/repeating requests from the admin server

In conclusion, this patch makes the system behave better in error cases, but it does not solve the problem if a file is corrupt.  But at least you will not get workers or server spinning in an infinite loop if a database server is restarted, which was possible without this patch.

**Comment by destijl** [comment]
LGTM. The downside of the sqlite datastore seems to be that it's fairly easy to corrupt one of the individual files. We should look at other ways we can make the problem more obvious, feel free to suggest any ideas you have.

**Comment by destijl** [comment]
Thanks for your work. Let me know how you'd like to be listed in the ACK file:
https://github.com/google/grr/blob/master/ACKNOWLEDGEMENTS
